[Bug]: The fullmajor optimize task commit failed

### What happened?

All optimize tasks of a table are in the "prepared" state. I checked the AMS log and found that the optimize task has been completed, but every commit fails. In addition, the optimizer reported a failover before, because the resource manager restarted.


### Affects Versions

0.3.2

### What engines are you seeing the problem on?

Optimizer

### How to reproduce

step 1: optimizer failover
step 2: start new optimize task

### Relevant log output

```shell
2022-11-18 09:39:50,633 ERROR [Metastore Scheduled Commit Worker 3] [com.netease.arctic.ams.server.optimize.OptimizeCommitWorker] [] - Optimize Committer Worker-3 com.netease.arctic.ams.server.optimize.TableOptimizeItem@7b863ac5 unexpected commit error
java.lang.Exception: unexpected commit error
        at com.netease.arctic.ams.server.optimize.BaseOptimizeCommit.commit(BaseOptimizeCommit.java:145) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.SupportHiveCommit.commit(SupportHiveCommit.java:110) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.TableOptimizeItem.commitOptimizeTasks(TableOptimizeItem.java:823) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.OptimizeCommitWorker.run(OptimizeCommitWorker.java:50) [arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_261]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_261]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_261]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_261]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_261]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_261]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_261]
Caused by: com.netease.arctic.hive.exceptions.CannotAlterHiveLocationException: can't delete hive partition: Partition(values: [2022-11-17/android], location: hdfs://eadhadoop/user/da_luna/hive_db/da_luna.db/arctic_dict_homepage_outlog/hive/day=2022-11-17/client=android/1668693508469_31637), file under partition is not deleted: hdfs://eadhadoop/user/da_luna/hive_db/da_luna.db/arctic_dict_homepage_outlog/hive/day=2022-11-17/client=android/1668693508469_31637/10-B-31637-00000-593788945-0000000001.parquet
        at com.netease.arctic.hive.op.UpdateHiveFiles.checkPartitionDelete(UpdateHiveFiles.java:238) ~[arctic-hive-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.hive.op.UpdateHiveFiles.lambda$getDeletePartition$5(UpdateHiveFiles.java:225) ~[arctic-hive-0.3.2-SNAPSHOT.jar:?]
        at java.util.HashMap$Values.forEach(HashMap.java:981) ~[?:1.8.0_261]
        at com.netease.arctic.hive.op.UpdateHiveFiles.getDeletePartition(UpdateHiveFiles.java:225) ~[arctic-hive-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.hive.op.UpdateHiveFiles.commit(UpdateHiveFiles.java:97) ~[arctic-hive-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.BaseOptimizeCommit.majorCommit(BaseOptimizeCommit.java:314) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.BaseOptimizeCommit.commit(BaseOptimizeCommit.java:129) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
```


### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/ams/ams-api/src/main/java/com/netease/arctic/ams/api/properties/OptimizeTaskProperties.java
+++ b/ams/ams-api/src/main/java/com/netease/arctic/ams/api/properties/OptimizeTaskProperties.java
@@ -4,4 +4,5 @@ public class OptimizeTaskProperties {
   // optimize task properties
   public static final String ALL_FILE_COUNT = "all-file-cnt";
   public static final String CUSTOM_HIVE_SUB_DIRECTORY = "custom-hive-sub-directory";
+  public static final String MAX_EXECUTE_TIME = "max-execute-time";
 }

--- a/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java
+++ b/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java
@@ -218,6 +218,8 @@ public class OptimizeTaskItem extends IJDBCService {
               .getTableOptimizeItem(getTableIdentifier()).getArcticTable();
           for (FileStatus fileStatus : arcticTable.io().list(location)) {
             String fileLocation = fileStatus.getPath().toUri().getPath();
+            // now file naming rule is nodeId-fileType-txId-partitionId-taskId-fileCount(%d-%s-%d-%05d-%d-%010d)
+            // for files produced by optimize, the taskId is attemptId
             String pattern = ".*(\\d{5}-)" + optimizeRuntime.getAttemptId() + "(-\\d{10}).*";
             if (Pattern.matches(pattern, fileLocation)) {
               arcticTable.io().deleteFile(fileLocation);

--- a/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java
+++ b/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java
@@ -51,6 +51,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.Supplier;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 public class OptimizeTaskItem extends IJDBCService {
@@ -217,7 +218,8 @@ public class OptimizeTaskItem extends IJDBCService {
               .getTableOptimizeItem(getTableIdentifier()).getArcticTable();
           for (FileStatus fileStatus : arcticTable.io().list(location)) {
             String fileLocation = fileStatus.getPath().toUri().getPath();
-            if (fileLocation.contains(optimizeRuntime.getAttemptId())) {
+            String pattern = ".*(\\d{5}-)" + optimizeRuntime.getAttemptId() + "(-\\d{10}).*";
+            if (Pattern.matches(pattern, fileLocation)) {
               arcticTable.io().deleteFile(fileLocation);
               LOG.debug("delete file {} by produced failed optimize task.", fileLocation);
             }

--- a/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java
+++ b/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java
@@ -94,18 +94,6 @@ public class OptimizeTaskItem extends IJDBCService {
     try {
       Preconditions.checkArgument(optimizeRuntime.getStatus() != OptimizeStatus.Prepared,
           "task prepared, can't on pending");
-
-      // can update max execute time on optimizing
-      try {
-        ArcticTable arcticTable = ServiceContainer.getOptimizeService()
-            .getTableOptimizeItem(getTableIdentifier()).getArcticTable();
-        Long maxExecuteTime = PropertyUtil.propertyAsLong(arcticTable.properties(),
-            TableProperties.OPTIMIZE_EXECUTE_TIMEOUT, TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT);
-        optimizeTask.getProperties().put(OptimizeTaskProperties.MAX_EXECUTE_TIME, String.valueOf(maxExecuteTime));
-      } catch (Exception e) {
-        LOG.error("update task max execute time failed.", e);
-      }
-
       BaseOptimizeTaskRuntime newRuntime = optimizeRuntime.clone();
       if (newRuntime.getStatus() == OptimizeStatus.Failed) {
         newRuntime.setRetry(newRuntime.getRetry() + 1);
@@ -210,28 +198,6 @@ public class OptimizeTaskItem extends IJDBCService {
     if (getOptimizeStatus() == OptimizeStatus.Init) {
       return true;
     } else if (getOptimizeStatus() == OptimizeStatus.Failed) {
-      // clear useless files produced by failed full optimize task support hive
-      String location = getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
-      if (location != null) {
-        try {
-          ArcticTable arcticTable = ServiceContainer.getOptimizeService()
-              .getTableOptimizeItem(getTableIdentifier()).getArcticTable();
-          for (FileStatus fileStatus : arcticTable.io().list(location)) {
-            String fileLocation = fileStatus.getPath().toUri().getPath();
-            // now file naming rule is nodeId-fileType-txId-partitionId-taskId-fileCount(%d-%s-%d-%05d-%d-%010d)
-            // for files produced by optimize, the taskId is attemptId
-            String pattern = ".*(\\d{5}-)" + optimizeRuntime.getAttemptId() + "(-\\d{10}).*";
-            if (Pattern.matches(pattern, fileLocation)) {
-              arcticTable.io().deleteFile(fileLocation);
-              LOG.debug("delete file {} by produced failed optimize task.", fileLocation);
-            }
-          }
-        } catch (Exception e) {
-          LOG.error("delete files failed", e);
-          return false;
-        }
-      }
-
       return getOptimizeRuntime().getRetry() <= maxRetry.get() && System.currentTimeMillis() >
           getOptimizeRuntime().getFailTime() + RETRY_INTERVAL;
     }
@@ -295,6 +261,19 @@ public class OptimizeTaskItem extends IJDBCService {
         selectOptimizeTaskFiles(DataFileType.POS_DELETE_FILE.name(), 1));
     optimizeRuntime.setTargetFiles(targetFiles.stream()
         .map(SerializationUtil::byteArrayToByteBuffer).collect(Collectors.toList()));
+  }
+
+  public void setMaxExecuteTime() {
+    // can update max execute time on optimizing
+    try {
+      ArcticTable arcticTable = ServiceContainer.getOptimizeService()
+          .getTableOptimizeItem(getTableIdentifier()).getArcticTable();
+      Long maxExecuteTime = PropertyUtil.propertyAsLong(arcticTable.properties(),
+          TableProperties.OPTIMIZE_EXECUTE_TIMEOUT, TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT);
+      optimizeTask.getProperties().put(OptimizeTaskProperties.MAX_EXECUTE_TIME, String.valueOf(maxExecuteTime));
+    } catch (Exception e) {
+      LOG.error("update task max execute time failed.", e);
+    }
   }
 
   private List<byte[]> selectOptimizeTaskFiles(String contentType, int isTarget) {

--- a/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java
+++ b/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java
@@ -23,6 +23,7 @@ import com.netease.arctic.ams.api.ErrorMessage;
 import com.netease.arctic.ams.api.JobId;
 import com.netease.arctic.ams.api.OptimizeStatus;
 import com.netease.arctic.ams.api.OptimizeTaskId;
+import com.netease.arctic.ams.api.properties.OptimizeTaskProperties;
 import com.netease.arctic.ams.server.mapper.InternalTableFilesMapper;
 import com.netease.arctic.ams.server.mapper.OptimizeTaskRuntimesMapper;
 import com.netease.arctic.ams.server.mapper.OptimizeTasksMapper;
@@ -31,12 +32,17 @@ import com.netease.arctic.ams.server.model.BaseOptimizeTask;
 import com.netease.arctic.ams.server.model.BaseOptimizeTaskRuntime;
 import com.netease.arctic.ams.server.model.TableTaskHistory;
 import com.netease.arctic.ams.server.service.IJDBCService;
+import com.netease.arctic.ams.server.service.ServiceContainer;
 import com.netease.arctic.data.DataFileType;
+import com.netease.arctic.table.ArcticTable;
 import com.netease.arctic.table.TableIdentifier;
+import com.netease.arctic.table.TableProperties;
 import com.netease.arctic.utils.SerializationUtil;
+import org.apache.hadoop.fs.FileStatus;
 import org.apache.ibatis.session.SqlSession;
 import org.apache.iceberg.ContentFile;
 import org.apache.iceberg.FileContent;
+import org.apache.iceberg.util.PropertyUtil;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -87,6 +93,18 @@ public class OptimizeTaskItem extends IJDBCService {
     try {
       Preconditions.checkArgument(optimizeRuntime.getStatus() != OptimizeStatus.Prepared,
           "task prepared, can't on pending");
+
+      // can update max execute time on optimizing
+      try {
+        ArcticTable arcticTable = ServiceContainer.getOptimizeService()
+            .getTableOptimizeItem(getTableIdentifier()).getArcticTable();
+        Long maxExecuteTime = PropertyUtil.propertyAsLong(arcticTable.properties(),
+            TableProperties.OPTIMIZE_EXECUTE_TIMEOUT, TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT);
+        optimizeTask.getProperties().put(OptimizeTaskProperties.MAX_EXECUTE_TIME, String.valueOf(maxExecuteTime));
+      } catch (Exception e) {
+        LOG.error("update task max execute time failed.", e);
+      }
+
       BaseOptimizeTaskRuntime newRuntime = optimizeRuntime.clone();
       if (newRuntime.getStatus() == OptimizeStatus.Failed) {
         newRuntime.setRetry(newRuntime.getRetry() + 1);
@@ -151,7 +169,6 @@ public class OptimizeTaskItem extends IJDBCService {
       newRuntime.setStatus(OptimizeStatus.Failed);
       newRuntime.setPreparedTime(BaseOptimizeTaskRuntime.INVALID_TIME);
       newRuntime.setReportTime(reportTime);
-      newRuntime.setAttemptId(null);
       newRuntime.setCostTime(costTime);
       persistTaskRuntime(newRuntime, false);
       optimizeRuntime = newRuntime;
@@ -192,15 +209,37 @@ public class OptimizeTaskItem extends IJDBCService {
     if (getOptimizeStatus() == OptimizeStatus.Init) {
       return true;
     } else if (getOptimizeStatus() == OptimizeStatus.Failed) {
+      // clear useless files produced by failed full optimize task support hive
+      String location = getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
+      if (location != null) {
+        try {
+          ArcticTable arcticTable = ServiceContainer.getOptimizeService()
+              .getTableOptimizeItem(getTableIdentifier()).getArcticTable();
+          for (FileStatus fileStatus : arcticTable.io().list(location)) {
+            String fileLocation = fileStatus.getPath().toUri().getPath();
+            if (fileLocation.contains(optimizeRuntime.getAttemptId())) {
+              arcticTable.io().deleteFile(fileLocation);
+              LOG.debug("delete file {} by produced failed optimize task.", fileLocation);
+            }
+          }
+        } catch (Exception e) {
+          LOG.error("delete files failed", e);
+          return false;
+        }
+      }
+
       return getOptimizeRuntime().getRetry() <= maxRetry.get() && System.currentTimeMillis() >
           getOptimizeRuntime().getFailTime() + RETRY_INTERVAL;
     }
     return false;
   }
 
-  public boolean executeTimeout(Supplier<Long> maxExecuteTime) {
+  public boolean executeTimeout() {
+    long maxExecuteTime = PropertyUtil
+        .propertyAsLong(optimizeTask.getProperties(), OptimizeTaskProperties.MAX_EXECUTE_TIME,
+            TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT);
     if (getOptimizeStatus() == OptimizeStatus.Executing) {
-      return System.currentTimeMillis() - optimizeRuntime.getExecuteTime() > maxExecuteTime.get();
+      return System.currentTimeMillis() - optimizeRuntime.getExecuteTime() > maxExecuteTime;
     }
     return false;
   }

--- a/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/TableOptimizeItem.java
+++ b/ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/TableOptimizeItem.java
@@ -761,18 +761,12 @@ public class TableOptimizeItem extends IJDBCService {
    * If task execute timeout, set it to be Failed.
    */
   public void checkTaskExecuteTimeout() {
-    optimizeTasks.values().stream().filter(task -> task.executeTimeout(this::maxExecuteTime))
+    optimizeTasks.values().stream().filter(task -> task.executeTimeout())
         .forEach(task -> {
           task.onFailed(new ErrorMessage(System.currentTimeMillis(), "execute expired"),
               System.currentTimeMillis() - task.getOptimizeRuntime().getExecuteTime());
           LOG.error("{} execute timeout, change to Failed", task.getTaskId());
         });
-  }
-
-  private long maxExecuteTime() {
-    return PropertyUtil
-        .propertyAsLong(getArcticTable(false).properties(), TableProperties.OPTIMIZE_EXECUTE_TIMEOUT,
-            TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT);
   }
 
   /**

--- a/ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java
+++ b/ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java
@@ -27,6 +27,7 @@ import com.netease.arctic.ams.api.NoSuchObjectException;
 import com.netease.arctic.ams.api.OptimizeStatus;
 import com.netease.arctic.ams.api.OptimizeTask;
 import com.netease.arctic.ams.api.OptimizeType;
+import com.netease.arctic.ams.api.properties.OptimizeTaskProperties;
 import com.netease.arctic.ams.server.mapper.ContainerMetadataMapper;
 import com.netease.arctic.ams.server.mapper.OptimizeQueueMapper;
 import com.netease.arctic.ams.server.model.BaseOptimizeTask;
@@ -51,6 +52,7 @@ import com.netease.arctic.table.TableProperties;
 import com.netease.arctic.utils.TableTypeUtil;
 import org.apache.commons.collections.CollectionUtils;
 import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.fs.FileStatus;
 import org.apache.ibatis.session.SqlSession;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 import org.apache.iceberg.util.PropertyUtil;
@@ -74,6 +76,7 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.Condition;
 import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 public class OptimizeQueueService extends IJDBCService {
@@ -374,6 +377,27 @@ public class OptimizeQueueService extends IJDBCService {
           throw new InvalidObjectException(
               queueName() + " not allow task of table " + task.getTableIdentifier());
         }
+        // clear useless files produced by failed full optimize task support hive
+        String location = task.getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
+        if (location != null) {
+          try {
+            ArcticTable arcticTable = ServiceContainer.getOptimizeService()
+                .getTableOptimizeItem(task.getTableIdentifier()).getArcticTable();
+            for (FileStatus fileStatus : arcticTable.io().list(location)) {
+              String fileLocation = fileStatus.getPath().toUri().getPath();
+              // now file naming rule is nodeId-fileType-txId-partitionId-taskId-fileCount(%d-%s-%d-%05d-%d-%010d)
+              // for files produced by optimize, the taskId is attemptId
+              String pattern = ".*(\\d{5}-)" + task.getOptimizeRuntime().getAttemptId() + "(-\\d{10}).*";
+              if (Pattern.matches(pattern, fileLocation)) {
+                arcticTable.io().deleteFile(fileLocation);
+                LOG.debug("delete file {} by produced failed optimize task.", fileLocation);
+              }
+            }
+          } catch (Exception e) {
+            LOG.error("delete files failed", e);
+            return;
+          }
+        }
         if (tasks.offer(task)) {
           if (!OptimizeStatusUtil.in(task.getOptimizeStatus(), OptimizeStatus.Pending) ||
               task.getOptimizeTask().getQueueId() != optimizeQueue.getOptimizeQueueMeta()
@@ -490,6 +514,8 @@ public class OptimizeQueueService extends IJDBCService {
                 task.onFailed(new ErrorMessage(System.currentTimeMillis(), "failed to put task back into queue"), 0);
               }
             }
+            // update max execute time
+            task.setMaxExecuteTime();
             TableTaskHistory tableTaskHistory = task.onExecuting(jobId, attemptId);
             try {
               insertTableTaskHistory(tableTaskHistory);

--- a/ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java
+++ b/ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java
@@ -379,7 +379,8 @@ public class OptimizeQueueService extends IJDBCService {
         }
         // clear useless files produced by failed full optimize task support hive
         if (task.getOptimizeRuntime().getStatus() == OptimizeStatus.Failed) {
-          String location = task.getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
+          String location =
+              task.getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
 
           if (location != null) {
             try {

--- a/ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java
+++ b/ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java
@@ -378,26 +378,32 @@ public class OptimizeQueueService extends IJDBCService {
               queueName() + " not allow task of table " + task.getTableIdentifier());
         }
         // clear useless files produced by failed full optimize task support hive
-        String location = task.getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
-        if (location != null) {
-          try {
-            ArcticTable arcticTable = ServiceContainer.getOptimizeService()
-                .getTableOptimizeItem(task.getTableIdentifier()).getArcticTable();
-            for (FileStatus fileStatus : arcticTable.io().list(location)) {
-              String fileLocation = fileStatus.getPath().toUri().getPath();
-              // now file naming rule is nodeId-fileType-txId-partitionId-taskId-fileCount(%d-%s-%d-%05d-%d-%010d)
-              // for files produced by optimize, the taskId is attemptId
-              String pattern = ".*(\\d{5}-)" + task.getOptimizeRuntime().getAttemptId() + "(-\\d{10}).*";
-              if (Pattern.matches(pattern, fileLocation)) {
-                arcticTable.io().deleteFile(fileLocation);
-                LOG.debug("delete file {} by produced failed optimize task.", fileLocation);
+        if (task.getOptimizeRuntime().getStatus() == OptimizeStatus.Failed) {
+          String location = task.getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
+
+          if (location != null) {
+            try {
+              ArcticTable arcticTable = ServiceContainer.getOptimizeService()
+                  .getTableOptimizeItem(task.getTableIdentifier()).getArcticTable();
+              if (arcticTable.io().exists(location)) {
+                for (FileStatus fileStatus : arcticTable.io().list(location)) {
+                  String fileLocation = fileStatus.getPath().toUri().getPath();
+                  // now file naming rule is nodeId-fileType-txId-partitionId-taskId-fileCount(%d-%s-%d-%05d-%d-%010d)
+                  // for files produced by optimize, the taskId is attemptId
+                  String pattern = ".*(\\d{5}-)" + task.getOptimizeRuntime().getAttemptId() + "(-\\d{10}).*";
+                  if (Pattern.matches(pattern, fileLocation)) {
+                    arcticTable.io().deleteFile(fileLocation);
+                    LOG.debug("delete file {} by produced failed optimize task.", fileLocation);
+                  }
+                }
               }
+            } catch (Exception e) {
+              LOG.error("delete files failed", e);
+              return;
             }
-          } catch (Exception e) {
-            LOG.error("delete files failed", e);
-            return;
           }
         }
+
         if (tasks.offer(task)) {
           if (!OptimizeStatusUtil.in(task.getOptimizeStatus(), OptimizeStatus.Pending) ||
               task.getOptimizeTask().getQueueId() != optimizeQueue.getOptimizeQueueMeta()

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/exception/TimeoutException.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/exception/TimeoutException.java
@@ -1,0 +1,25 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.optimizer.exception;
+
+public class TimeoutException extends RuntimeException {
+  public TimeoutException(String errorMessage) {
+    super(errorMessage);
+  }
+}

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/BaseTaskExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/BaseTaskExecutor.java
@@ -34,18 +34,21 @@ import com.netease.arctic.data.DataFileType;
 import com.netease.arctic.data.DataTreeNode;
 import com.netease.arctic.optimizer.OptimizerConfig;
 import com.netease.arctic.optimizer.TaskWrapper;
+import com.netease.arctic.optimizer.exception.TimeoutException;
 import com.netease.arctic.optimizer.operator.executor.Executor;
 import com.netease.arctic.optimizer.operator.executor.ExecutorFactory;
 import com.netease.arctic.optimizer.operator.executor.NodeTask;
 import com.netease.arctic.optimizer.operator.executor.OptimizeTaskResult;
 import com.netease.arctic.optimizer.operator.executor.TableIdentificationInfo;
 import com.netease.arctic.table.ArcticTable;
+import com.netease.arctic.table.TableProperties;
 import com.netease.arctic.utils.SerializationUtil;
 import org.apache.commons.beanutils.BeanUtils;
 import org.apache.commons.collections.CollectionUtils;
 import org.apache.iceberg.ContentFile;
 import org.apache.iceberg.FileScanTask;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.PropertyUtil;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -126,6 +129,10 @@ public class BaseTaskExecutor implements Serializable {
       OptimizeTaskResult<?> result = optimize.execute();
       onTaskFinish(result.getTargetFiles());
       return result.getOptimizeTaskStat();
+    } catch (TimeoutException timeoutException) {
+      LOG.error("execute task timeout {}", task.getTaskId());
+      onTaskFailed(timeoutException);
+      return constructFailedResult(task, timeoutException);
     } catch (Throwable t) {
       LOG.error("failed to execute task {}", task.getTaskId(), t);
       onTaskFailed(t);
@@ -279,6 +286,10 @@ public class BaseTaskExecutor implements Serializable {
 
       String customHiveSubdirectory = properties.get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
       nodeTask.setCustomHiveSubdirectory(customHiveSubdirectory);
+
+      Long maxExecuteTime = PropertyUtil.propertyAsLong(properties,
+          OptimizeTaskProperties.MAX_EXECUTE_TIME, TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT);
+      nodeTask.setMaxExecuteTime(maxExecuteTime);
     }
 
     return nodeTask;

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java
@@ -19,11 +19,18 @@
 package com.netease.arctic.optimizer.operator.executor;
 
 import com.netease.arctic.data.DataTreeNode;
+import com.netease.arctic.optimizer.OptimizerConfig;
+import com.netease.arctic.optimizer.exception.TimeoutException;
+import com.netease.arctic.table.ArcticTable;
+import com.netease.arctic.table.TableProperties;
 import com.netease.arctic.utils.FileUtil;
 import org.apache.iceberg.ContentFile;
 import org.apache.iceberg.DataFile;
 import org.apache.iceberg.DeleteFile;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import java.io.Closeable;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -31,6 +38,24 @@ import java.util.OptionalLong;
 import java.util.stream.Collectors;
 
 public abstract class BaseExecutor<F extends ContentFile<F>> implements Executor<F> {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseExecutor.class);
+
+  protected final NodeTask task;
+  protected final ArcticTable table;
+  protected final OptimizerConfig config;
+  protected final long startTime;
+  protected double factor = 0.9;
+
+  public BaseExecutor(NodeTask nodeTask,
+                      ArcticTable table,
+                      long startTime,
+                      OptimizerConfig config) {
+    this.startTime = startTime;
+    this.task = nodeTask;
+    this.table = table;
+    this.config = config;
+  }
+
   protected Map<DataTreeNode, List<DataFile>> groupDataFilesByNode(List<DataFile> dataFiles) {
     return new HashMap<>(dataFiles.stream().collect(Collectors.groupingBy(dataFile ->
         FileUtil.parseFileNodeFromFileName(dataFile.path().toString()))));
@@ -49,5 +74,16 @@ public abstract class BaseExecutor<F extends ContentFile<F>> implements Executor
     }
 
     return 0;
+  }
+
+  protected void executeTimeout(Closeable writer) throws Exception {
+    long maxExecuteTime = task.getMaxExecuteTime() != null ?
+        task.getMaxExecuteTime() : TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT;
+    if (System.currentTimeMillis() - startTime > maxExecuteTime * factor) {
+      writer.close();
+      LOG.info("table {} execute task {} timeout, max execute time is {}",
+          table.id(), task.getTaskId(), task.getMaxExecuteTime());
+      throw new TimeoutException("optimizer execute timeout");
+    }
   }
 }

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java
@@ -79,11 +79,14 @@ public abstract class BaseExecutor<F extends ContentFile<F>> implements Executor
   protected void checkIfTimeout(Closeable writer) throws Exception {
     long maxExecuteTime = task.getMaxExecuteTime() != null ?
         task.getMaxExecuteTime() : TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT;
-    if (System.currentTimeMillis() - startTime > maxExecuteTime * factor) {
+    long actualExecuteTime = System.currentTimeMillis() - startTime;
+    if (actualExecuteTime > maxExecuteTime * factor) {
       writer.close();
-      LOG.info("table {} execute task {} timeout, max execute time is {}",
-          table.id(), task.getTaskId(), task.getMaxExecuteTime());
-      throw new TimeoutException("optimizer execute timeout");
+      LOG.error("table {} execute task {} timeout, actual execute time is {}ms, max execute time is {}ms",
+          table.id(), task.getTaskId(), actualExecuteTime, task.getMaxExecuteTime());
+      throw new TimeoutException(String.format("optimizer execute timeout, " +
+          "actual execute time is %sms, max execute time is %sms, factor is %s",
+          actualExecuteTime, task.getMaxExecuteTime(), factor));
     }
   }
 }

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java
@@ -76,7 +76,7 @@ public abstract class BaseExecutor<F extends ContentFile<F>> implements Executor
     return 0;
   }
 
-  protected void executeTimeout(Closeable writer) throws Exception {
+  protected void checkIfTimeout(Closeable writer) throws Exception {
     long maxExecuteTime = task.getMaxExecuteTime() != null ?
         task.getMaxExecuteTime() : TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT;
     if (System.currentTimeMillis() - startTime > maxExecuteTime * factor) {

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/IcebergExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/IcebergExecutor.java
@@ -124,7 +124,7 @@ public class IcebergExecutor extends BaseExecutor<DataFile> {
 
     long insertCount = 0;
     while (recordIterator.hasNext()) {
-      executeTimeout(writer);
+      checkIfTimeout(writer);
 
       if (writer.length() >= targetSizeByBytes) {
         writer.close();

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/IcebergExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/IcebergExecutor.java
@@ -55,19 +55,11 @@ import static org.apache.iceberg.TableProperties.DEFAULT_FILE_FORMAT_DEFAULT;
 public class IcebergExecutor extends BaseExecutor<DataFile> {
   private static final Logger LOG = LoggerFactory.getLogger(IcebergExecutor.class);
 
-  private final NodeTask task;
-  private final ArcticTable table;
-  private final long startTime;
-  private final OptimizerConfig config;
-
   public IcebergExecutor(NodeTask nodeTask,
                          ArcticTable table,
                          long startTime,
                          OptimizerConfig config) {
-    this.task = nodeTask;
-    this.table = table;
-    this.startTime = startTime;
-    this.config = config;
+    super(nodeTask, table, startTime, config);
   }
 
   @Override
@@ -132,6 +124,8 @@ public class IcebergExecutor extends BaseExecutor<DataFile> {
 
     long insertCount = 0;
     while (recordIterator.hasNext()) {
+      executeTimeout(writer);
+
       if (writer.length() >= targetSizeByBytes) {
         writer.close();
         result.add(writer.toDataFile());

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/MajorExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/MajorExecutor.java
@@ -131,7 +131,7 @@ public class MajorExecutor extends BaseExecutor<DataFile> {
             WriteOperationKind.MAJOR_OPTIMIZE : WriteOperationKind.FULL_OPTIMIZE);
     long insertCount = 0;
     while (recordIterator.hasNext()) {
-      executeTimeout(writer);
+      checkIfTimeout(writer);
 
       Record baseRecord = recordIterator.next();
       writer.write(baseRecord);

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/MajorExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/MajorExecutor.java
@@ -62,16 +62,11 @@ import java.util.stream.Collectors;
 public class MajorExecutor extends BaseExecutor<DataFile> {
   private static final Logger LOG = LoggerFactory.getLogger(MajorExecutor.class);
 
-  private final NodeTask task;
-  private final ArcticTable table;
-  private final long startTime;
-  private final OptimizerConfig config;
-
-  public MajorExecutor(NodeTask nodeTask, ArcticTable table, long startTime, OptimizerConfig config) {
-    this.task = nodeTask;
-    this.table = table;
-    this.startTime = startTime;
-    this.config = config;
+  public MajorExecutor(NodeTask nodeTask,
+                       ArcticTable table,
+                       long startTime,
+                       OptimizerConfig config) {
+    super(nodeTask, table, startTime, config);
   }
 
   @Override
@@ -136,6 +131,8 @@ public class MajorExecutor extends BaseExecutor<DataFile> {
             WriteOperationKind.MAJOR_OPTIMIZE : WriteOperationKind.FULL_OPTIMIZE);
     long insertCount = 0;
     while (recordIterator.hasNext()) {
+      executeTimeout(writer);
+
       Record baseRecord = recordIterator.next();
       writer.write(baseRecord);
       insertCount++;

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/MinorExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/MinorExecutor.java
@@ -63,16 +63,11 @@ import java.util.stream.Collectors;
 public class MinorExecutor extends BaseExecutor<DeleteFile> {
   private static final Logger LOG = LoggerFactory.getLogger(MinorExecutor.class);
 
-  private final NodeTask task;
-  private final ArcticTable table;
-  private final long startTime;
-  private final OptimizerConfig config;
-
-  public MinorExecutor(NodeTask nodeTask, ArcticTable table, long startTime, OptimizerConfig config) {
-    this.task = nodeTask;
-    this.table = table;
-    this.startTime = startTime;
-    this.config = config;
+  public MinorExecutor(NodeTask nodeTask,
+                       ArcticTable table,
+                       long startTime,
+                       OptimizerConfig config) {
+    super(nodeTask, table, startTime, config);
   }
 
   @Override
@@ -103,6 +98,8 @@ public class MinorExecutor extends BaseExecutor<DeleteFile> {
             openTask(dataFiles, posDeleteList, requiredSchema, task.getSourceNodes());
 
         while (iterator.hasNext()) {
+          executeTimeout(posDeleteWriter);
+
           Record record = iterator.next();
           String filePath = (String) record.get(recordStruct.fields()
               .indexOf(recordStruct.field(MetadataColumns.FILE_PATH.name())));
@@ -125,6 +122,8 @@ public class MinorExecutor extends BaseExecutor<DeleteFile> {
           CloseableIterable<Record> posDeleteIterable = posDeleteReader.readDeletes();
           CloseableIterator<Record> posDeleteIterator = posDeleteIterable.iterator();
           while (posDeleteIterator.hasNext()) {
+            executeTimeout(posDeleteWriter);
+
             Record record = posDeleteIterator.next();
             String filePath = posDeleteReader.readPath(record);
             Long rowPosition = posDeleteReader.readPos(record);

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/MinorExecutor.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/MinorExecutor.java
@@ -98,7 +98,7 @@ public class MinorExecutor extends BaseExecutor<DeleteFile> {
             openTask(dataFiles, posDeleteList, requiredSchema, task.getSourceNodes());
 
         while (iterator.hasNext()) {
-          executeTimeout(posDeleteWriter);
+          checkIfTimeout(posDeleteWriter);
 
           Record record = iterator.next();
           String filePath = (String) record.get(recordStruct.fields()
@@ -122,7 +122,7 @@ public class MinorExecutor extends BaseExecutor<DeleteFile> {
           CloseableIterable<Record> posDeleteIterable = posDeleteReader.readDeletes();
           CloseableIterator<Record> posDeleteIterator = posDeleteIterable.iterator();
           while (posDeleteIterator.hasNext()) {
-            executeTimeout(posDeleteWriter);
+            checkIfTimeout(posDeleteWriter);
 
             Record record = posDeleteIterator.next();
             String filePath = posDeleteReader.readPath(record);

--- a/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/NodeTask.java
+++ b/optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/NodeTask.java
@@ -52,6 +52,7 @@ public class NodeTask {
   private TableIdentifier tableIdentifier;
   private int attemptId;
   private String customHiveSubdirectory;
+  private Long maxExecuteTime;
 
   public NodeTask() {
   }
@@ -165,6 +166,14 @@ public class NodeTask {
     this.customHiveSubdirectory = customHiveSubdirectory;
   }
 
+  public Long getMaxExecuteTime() {
+    return maxExecuteTime;
+  }
+
+  public void setMaxExecuteTime(Long maxExecuteTime) {
+    this.maxExecuteTime = maxExecuteTime;
+  }
+
   public OptimizeType getOptimizeType() {
     return taskId.getType();
   }
@@ -185,6 +194,7 @@ public class NodeTask {
         .add("posDeleteFiles", posDeleteFiles.size())
         .add("fileScanTasks", fileScanTasks.size())
         .add("customHiveSubdirectory", customHiveSubdirectory)
+        .add("maxExecuteTime", maxExecuteTime)
         .toString();
   }
 }

--- a/optimizer/src/test/java/com/netease/arctic/optimizer/operator/executor/TestIcebergExecutor.java
+++ b/optimizer/src/test/java/com/netease/arctic/optimizer/operator/executor/TestIcebergExecutor.java
@@ -11,6 +11,7 @@ import com.netease.arctic.ams.api.properties.CatalogMetaProperties;
 import com.netease.arctic.catalog.ArcticCatalog;
 import com.netease.arctic.catalog.CatalogLoader;
 import com.netease.arctic.optimizer.OptimizerConfig;
+import com.netease.arctic.optimizer.exception.TimeoutException;
 import com.netease.arctic.table.ArcticTable;
 import com.netease.arctic.table.UnkeyedTable;
 import org.apache.commons.collections.IteratorUtils;
@@ -173,6 +174,33 @@ public class TestIcebergExecutor {
     result.getTargetFiles().forEach(dataFile -> totalRecordCount.set(totalRecordCount.get() + dataFile.recordCount()));
 
     Assert.assertEquals(46, totalRecordCount.get());
+  }
+
+  @Test
+  public void testExecuteTimeout() throws Exception {
+    icebergTable.asUnkeyedTable().updateProperties()
+        .set(com.netease.arctic.table.TableProperties.OPTIMIZE_SMALL_FILE_SIZE_BYTES_THRESHOLD, "1000")
+        .set(TableProperties.WRITE_TARGET_FILE_SIZE_BYTES, "100")
+        .commit();
+    List<DataFile> dataFiles = insertDataFiles(icebergTable.asUnkeyedTable());
+    insertEqDeleteFiles(icebergTable.asUnkeyedTable());
+    insertPosDeleteFiles(icebergTable.asUnkeyedTable(), dataFiles);
+
+    NodeTask nodeTask = constructNodeTask(icebergTable.asUnkeyedTable());
+    nodeTask.setMaxExecuteTime(0L);
+    String[] arg = new String[0];
+    OptimizerConfig optimizerConfig = new OptimizerConfig(arg);
+    optimizerConfig.setOptimizerId("UnitTest");
+    IcebergExecutor
+        icebergExecutor = new IcebergExecutor(nodeTask, icebergTable, System.currentTimeMillis(), optimizerConfig);
+    boolean isTimeout = false;
+    try {
+      icebergExecutor.execute();
+    } catch (TimeoutException e) {
+      isTimeout = true;
+    }
+
+    Assert.assertTrue(isTimeout);
   }
 
   protected List<DataFile> insertDataFiles(UnkeyedTable arcticTable) throws IOException {

